### PR TITLE
[java] 1862 - Message Digest calls should be synchronized

### LINKED ISSUE #1862: [java] New rule for MessageDigest.getInstance
```java
static MessageDigest INSTANCE = MessageDigest.getInstance("SHA-512");

INSTANCE.digest(...);
```

This is not thread-safe code. That is, one should not cache the `MessageDigest.getInstance` instance and reuse it multiple times between threads. Similar to the [UnsynchronizedStaticFormatter](https://pmd.github.io/pmd-6.14.0/pmd_rules_java_multithreading.html#unsynchronizedstaticformatter) check.

I was bit by this and that's why I'm suggesting it as a new check - it might help someone. :)

### PR DESCRIPTION
**Pull request by AnthonyKot**
Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**
I extended UnsynchronizedStaticFormatterRule which implements suitable checks.

Fixes #1862

### CODE REVIEW

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/multithreading/xml/UnsynchronizedMessageDigestFormatter.xml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import java.security.MessageDigest;
+
+public class Foo {
+    private static final MessageDigest INSTANCE = MessageDigest.getInstance("SHA-512");
+    void bar() {
+        synchronized(INSTANCE) {
+            byte[] input = new byte[13];
+            INSTANCE.digest(input); // no thread-safety, protected by no static monitor
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/multithreading/xml/UnsynchronizedMessageDigestFormatter.xml]: this one is actually thread-safe, you expect 0 violations

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/multithreading/UnsynchronizedStaticMessageDigestRule.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+package net.sourceforge.pmd.lang.java.rule.multithreading;
+
+import java.security.MessageDigest;
+
+/**
+ * Using a MessageDigest which is static can cause
+ * unexpected results when used in a multi-threaded environment. This rule will
+ * find static MessageDigest which are used in an unsynchronized manner.
+ */
+public class UnsynchronizedStaticMessageDigestRule extends UnsynchronizedStaticFormatterRule {
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/multithreading/UnsynchronizedStaticMessageDigestRule.java]: having this extend `UnsynchronizedStaticFormatterRule` seems semantically incorrect even if it works… we should probably refactor `UnsynchronizedStaticFormatterRule` into a `UnsynchronizedStaticAccessRule` which can be configured through properties to track unsynchronized static access to any given types (with proper defaults to include `Formatter` and `MessageDigest`).

The `UnsynchronizedStaticFormatterRule` can be marked as deprecated at the ruleset level, point to the new implementation, overriding the property defaults to keep just looking for `Formatter`.

Moreover, the `THREAD_SAFE_FORMATTER` static should probably be replaced with a property of thread-safe classes to ignore, once again with proper defaults.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/multithreading/xml/UnsynchronizedMessageDigestFormatter.xml (hunk 1 above)
Comment (AnthonyKot) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/multithreading/xml/UnsynchronizedMessageDigestFormatter.xml]: I think so. Just wrongly copy-pasted comment

**Comment by adangel** [comment]
I'm not sure whether we should propose using synchronized in that case.
I don't know, what the use case is of sharing a `MessageDigest` instance... IMHO, we should flag, if someone defines a field of this type, since I think, sharing a MessageDigest instance is not necessary.
See also this stackoverflow question: https://stackoverflow.com/questions/17554998/need-thread-safe-messagedigest-in-java

Suggesting to synchronize over the instance would workaround the issue, that the MessageDigest instance should not be shared from the start. It's also a lightweight object, so we don't get a real benefit of reusing it.

**Comment by AnthonyKot** [comment]
Sounds reasonable. Should I create new PR with rule that checks MessageDigest::digest is called once per MessageDigest object declared?

**Comment by AnthonyKot** [comment]
And other topic Travis build results. It fails 50/50 with CI and locally with message:
"[ERROR] warning: runtime JAR files in the classpath have the version 1.2, which is older than the API version 1.3. Consider using the runtime of version 1.3, or pass '-api-version 1.2' explicitly to restrict the available APIs to the runtime of version 1.2. You can also pass '-language-version 1.2' instead, which will restrict not only the APIs to the specified version, but also the language features"
Do we have PR to fix it? I think upgrading to Kotlin 1.3 should be sufficiently.

**Comment by adangel** [comment]
> Sounds reasonable. Should I create new PR with rule that checks MessageDigest::digest is called 
> once per MessageDigest object declared?

Actually, calling multiple times `digest()` is totally valid. This is not the problem. Example usage: https://gist.github.com/adangel/1b408a871b09c1a4c16cd9e5f339c70f#file-messagedigesttest-java-L27 You can reuse the message digest, as long as it is within one method. Since the [digest()](https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/security/MessageDigest.html#digest()) method resets the digest, you can directly start calculating the next hash.

The problem starts, if the MessageDigest instance is shared between multiple threads, e.g. as a field.

Example code: https://gist.github.com/adangel/1b408a871b09c1a4c16cd9e5f339c70f#file-messagedigesttest-java-L47

The main method calculates two hashes with a shared MessageDigest instance in multiple threads multiple times and collects the results into a Set. If it worked, there should be only one result, because the calculated hash should be always the same. But you get multiple hashes and probably not once the correct result...

```
---
Expected Hash1: 6y6m3hp6rCaTxRIBlKAayi+jRitqJ7EVgvT9MUpYCsc=
Hash1 Results: [/74Yrn99F3KcpzPNBsLq7rXeF7+kurUhCLN63+FC1vk=, QS7c19Ga6bkekiDIjNQJJbxPU6fB8GIqg52+HVmE/jI=, gX+1v5ObfiXN7vCDIBIkcA+CGM9rGEZtiez6MubHpYo=, NmAaCS7yiJ6FCDn+BozZi0tpJaQLkKq8WplfQAmz2Co=, o4LYAvrFUUictQZLJjV7ZXb/uuNTe3WEPHzI5Xomy98=, es6+PM0YKGDe9QCkYeEFZ5xO93qaOJ3mo7Gv6LlWcy4=, 6hmAru2KhMxno2ffbju4VvK89hodV1sgMsMwv6m5NRY=, tpYDGkncmHKWd+Pelki+6FfXwy4Yrl3RA2dDlqtYv5I=, UHq3Q5Pghiv8GnwmdZq4R6ggV6lZuIly0AI6jGnuri8=, agnmZ7tnroU8bvNypU/1OlEOUn+bBWiMH4PZq1vgzRk=, xARU7VizierArTLE26poKmqunQSmxkjXlWWZX9V4gcg=, +LR80ZramGYD9NT/aGenhrCPsQ2aP5FyP9CMYKVEyik=, gBVoKPE001nVv+UDfttiKxQx3G779VLpQVUsv1SrkZU=, 7iGHPkuYhhGavGVhd7ZhKDV7j1X5R/aZfWmm5BRLt7o=, +nz+Uh5aF1cgLS47jWx9hWBh2SQo38nbU/IB5RLXhF8=, AcstOno9BI6c3xtp+jvUjrz/qof2u1lHxJ9L5C9VzwA=, jwxiYnRxqgSB21leWQFAHhswSQkCX2ytLqL4ZgkT0aA=, 2hJEiVjPgVFtW7sJSpjCBLSlFoxEx7dqRmFXYOZ+hLw=, 2laYvhe5tGliM1eZd5++yozl1JHA0mJDuv756hg3qdg=, yn4CCJ63UlxWa0YlEhTLua9cyj9nZQ9sxPy9c/JNDfc=, 6y6m3hp6rCaTxRIBlKAayi+jRitqJ7EVgvT9MUpYCsc=, POtbPGnKu8vNhzYd5qTDhw/zoKx6L3DP20ZP/u3FFTs=, jiRbRyn0UyTZW4hYJ151GXAb34X7XfIUZjMPz/3PfY8=, sGOiA/CEGFW3jeGzRCPYJA76BsxApwwucLsLlYRwoEk=, h18TO3pNgkfLeraMK39Z0hARD5uZebpYYB0/Fe6oxzk=, /Z28L1fEPWWLycUxHBUL0FfMZrNhkf0MAQYBfMSmgCQ=, wEiUUX57eHQ23wOxI96MFf/cQidTpkrSeOcn5JAKhvs=, ayhwt5qYGcxZske/p5tLywiB4Y5pn1VO6BeNePHc7RQ=, 1fnAcY9STlt9i8/kzJ2Z2/SifUqY0bowQcTpzgM0xjY=, T1ry5+ggNtFRTbg6Vw1KTJ5h+Rcv08DXeuwK2PcsPho=, 72w3jMkqHlEsCfj07tubi3BlJqJIpnEhuq4sC6vWMbg=, 47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=, JWs0/rejgeG45HFBDs7KUezOjcVe6yKog/h3c+jjPAg=]
---
Expected Hash2: +LR80ZramGYD9NT/aGenhrCPsQ2aP5FyP9CMYKVEyik=
Hash2 Results: [MLs+xHSjXbiEdTbTj3dCDYGhk2i/lfMU0eiqdcTxRKQ=, LAEbkYJhr2xG0sb1kOHaBwqWFZ2kInuwZp7OHY5D85Q=, Ar5d6sXbrNdXVDAnc9a8cJOb1PlZgFhqQLLff7JicQY=, bPzGx+yFiHBDZkbSiGY7mFbmm6kqJa4QGfiDXiXUegw=, oZp8wORN7c8Pvxk+irEZD+RT9JUUEoN1X0LoY/GrLK0=, 0y8pc+p2C9dR6oJ9bz3exn9VLAtjEZdz1OohI2kNXWk=, Gr2MfGlTWbg46ICUaGenhoNfcs+aFPssdMQ39aVEyik=, ehTQMopVsWRI6Cg04l0F4EPLOMNPVk4+CUoa28Bgzzo=, ew5qPGB3quR15oOaJsUXRg6C9V3IBOXF8fudyADx0JM=, agnmZ7tnroU8bvNypU/1OlEOUn+bBWiMH4PZq1vgzRk=, BMpijVBI9Iy5DcLXs68A6J7Qp7yoRzJgOM1i50MHnYM=, QVSMYPqTPlwFsquqMMKmbFM4PgcHN8dXrUeekJzlKH8=, teFx1Hni0Tyi2a8zpkry2ziDS2KIlGM21Bt+scFH2Gk=, +LR80ZramGYD9NT/aGenhrCPsQ2aP5FyP9CMYKVEyik=, lD8YOMtuOVBGh6wUHIQoEXGfYJ8s3MqK8OJPPIBUgxI=, ooR/Uer5ESzLDki1njrPLCGQDuKS3t2RU9lq/lky2DE=, I86k0ck9AhAjMQJdU5LnMw0TnokE3XYzd+lwA2472Bw=, QzxDr+aLsi+9Hvq/f2PLQWrC16/vfkQ1lXb0TQMv4Pw=, IoybMc6ofKfuSDBzyhS3bbIUGeeYO7hxd4QqG5w1Kv4=, 9kuH+AUtTeOosWlT967V9e0bhryXi0g4ajOXl2/0NXk=, bzrk8wN+hwykNY0N71pUfXMVlSlxZSoZ3Ikb/qCkMak=, Vdf+mdwhY2aDclo4jQEDkxVmpXxsWuokd47XaLWJeMA=, lj2oDRGHwjOV7wdxh8vnrDfPdeGWU8WmGCfywJ/Dpro=, MgQ2AosJFkOC032t9hn0jW40N33W3T4fXIzhDh0nnrw=, h18TO3pNgkfLeraMK39Z0hARD5uZebpYYB0/Fe6oxzk=, GVpDothhxhsQYxdK+X0tyX50weuqpZUpq8OAlLwUxfs=, M61ry4lpSRooNi7G7R7/qywwMy3THxgyn1kOOBzfPdM=, jtVf2jxcx7xoAIagkaau2ehCasc4RbiJj8iD7M+wmuA=, q42wc9Z4ARPnqA4cQxUrthlm97fO0NLu7QVIzWdWlkE=, pqSUU8rJ2OOhwIjfI7eEnZ5b/9pSAZmr76EAeRq/Cbc=, 47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=, AMJbtXLVfkuu4mgzW0q4iZtUmgvYKXXa5JxPAoPfjXs=, XHa/n7sOD/SlVd1Ng4aA3XnhSv9mSNNxmwIzHyZGvVw=]
---
```

If you switch here https://gist.github.com/adangel/1b408a871b09c1a4c16cd9e5f339c70f#file-messagedigesttest-java-L65-L66 from method `calculateHashShared` to `calculateHashShared`, you'll get the correct result:

```
---
Expected Hash1: 6y6m3hp6rCaTxRIBlKAayi+jRitqJ7EVgvT9MUpYCsc=
Hash1 Results: [6y6m3hp6rCaTxRIBlKAayi+jRitqJ7EVgvT9MUpYCsc=]
---
Expected Hash2: +LR80ZramGYD9NT/aGenhrCPsQ2aP5FyP9CMYKVEyik=
Hash2 Results: [+LR80ZramGYD9NT/aGenhrCPsQ2aP5FyP9CMYKVEyik=]
---
```

You could of course synchronize on the shared message digest, this will also compute the correct results then. But why would you do this? It simpler, just to not share the message digest instance. 

> And other topic Travis build results. It fails 50/50 with CI and locally with message:
> "[ERROR] warning: runtime JAR files in the classpath have the version 1.2, which is older than 
> the API version 1.3. Consider using the runtime of version 1.3, or pass '-api-version 1.2' explicitly 
> to restrict the available APIs to the runtime of version 1.2. You can also pass '-language-version 
> 1.2' instead, which will restrict not only the APIs to the specified version, but also the language 
> features"
> Do we have PR to fix it? I think upgrading to Kotlin 1.3 should be sufficiently.

That's for sure another PR. But it is not the reason why this PR fails. You can look at the build logs:
https://travis-ci.org/pmd/pmd/jobs/568379671#L3619 (Maybe you need to scroll to the end)

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (checkstyle-check) on project pmd-java: You have 1 Checkstyle violation. -> [Help 1]
```

And a couple of lines above:

```
[INFO] --- maven-checkstyle-plugin:3.0.0:check (checkstyle-check) @ pmd-java ---
[INFO] There is 1 error reported by Checkstyle 8.18 with /net/sourceforge/pmd/pmd-checkstyle-config.xml ruleset.
[ERROR] src/main/java/net/sourceforge/pmd/lang/java/rule/multithreading/UnsynchronizedStaticMessageDigestRule.java:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
```

So, in short: IMHO we should detect Field declarations of the type MessageDigest.

**Comment by AnthonyKot** [comment]
I made new independent PR:
https://github.com/pmd/pmd/pull/1971
